### ISSUE TITLE
Code correctly returning a 401 unauthorised error in RC2 now incorrectly returns a 500 error

### ISSUE THREAD

**Issue by dumbledad** [aspnet/Security#927]
We have an ASP.Net Core web service using Azure AD B2C for bearer token authentication. Our iOS client makes an API request and if the ASP.Net Core server returns a 401 Unauthorised error with the term "SecurityTokenExpiredException" in the body then the client refreshes its token and tries again.

This works great under ASP.Net Core RC2 but now that we have moved to the release version we cannot deploy our server without breaking our clients because the server no longer correctly returns a 401 Unauthorised error and instead returns a 500 error.

The code that sets up the 401 is in Startup.cs

```csharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.Use(async (context, next) =>
    {
        try
        {
            await next();
        }
        catch (Exception ex)
        {
            if (context.Response.HasStarted)
            {
                throw;
            }

            int statusCode = 500;

            if (ex is SecurityTokenInvalidSignatureException ||
                ex is SecurityTokenInvalidLifetimeException ||
                ex is SecurityTokenExpiredException)
            {
                statusCode = 401;
            }
            else if (ex is ResourceNotAuthorizedException)
            {
                statusCode = 403;
            }
            else if (ex is ResourceNotFoundException || ex is UserDoesNotExistException)
            {
                statusCode = 404;
            }

            context.Response.StatusCode = statusCode;
            await context.Response.WriteAsync(new ExceptionResponse(statusCode, ex).ToString());
        }
    });
   //[...]
```

(N.B. If I step through this code in the debugger each line executes, the local variable `statusCode` is correctly set to 401, and those final two lines are run, but the error sent back by ASP.Net Core to the client is 500 and not 401.)

Has something changed going from RC2 to the ASP.Net Core release that would render this code incorrect? If so what is the new way of writing this?

(I have cross-posted this as a Stack Overflow question too [here](http://stackoverflow.com/q/38637454/575530). I hope that's OK.)

**Comment by dumbledad** [comment]
I have created a Gist showing the stark difference between the text written to `context.Response.WriteAsync` and the response text received on the client. The Gist is [here](https://gist.github.com/dumbledad/9d3f70b5621eb004e5f8758c4dcdc3f4)

**Comment by Eilon** [comment]
@troydai can you debug into this and see what's going on here? The SO question might have some more hints.

**Comment by troydai** [comment]
maybe related: https://github.com/aspnet/Security/issues/699

**Comment by troydai** [comment]
I found the root cause and I will work on fixes after I finish writing my findings.

### The problematic flow
1. Throw exception in the _AuthenticationFailed_ event trigger the flawed work flow. 
2. The _AuthenticationFailed_ event is called twice in `HandleAuthenticateAsync` of `JwtBearerHandler`: [Here](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L156) and [Here](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L176). The arrangement is flawed. The first invocation is in the _try_ clause, and the second is in the _catch_, therefore the event will be invoked twice. And for the second time it is invoked exception will be thrown from the _catch_ clause, which is oddly but clearly by design since any exception will be intentionally thrown [here](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L179).
3. The exception does interrupt the inner middleware and bust the flow into the outer middleware, where response status code is correctly set to 401. But just before the response writing headers, [the method registered](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs#L100) with response's _OnStarting_ event is invoked.
4. The _on starting_ flow eventually leads to method `HandleAuthomaticChallengeIfNeed`, where the status code __401__ is treated [differently](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs#L150).
5. The real issue happens in `JwtBearerHandler`'s `HandleUnauthorizeAsync` method. The exception is not expected from [retrieving task result from `HandleAuthenticateOnceAsync`](https://github.com/aspnet/Security/blob/c16fc06cb9fd28c21556f83855498985739e41e8/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L188). Since the exception is thrown earlier in `HandleAuthenticateAsync`, so the task it returns still trapped the exception. 

### Issues
1. `JwtBearerHandler:HandleUnauthorizationAsync` should handle the exception thrown from `HandleAuthenticateOnceAsync`.
2. Repeatedly execute _AuthenticateFailed_ event under failed authenticating flow is wrong.

### Other issues
It doesn't seem to me that the code is well prepared for exception thrown from events since they're called in the _catch_ clause. We should have a clear guideline to application developer about what should be done in the event and also better prepare for fatal actions in these extensions point.

Additionally, @dumbledad using exception to control the flow will be always tricky since it taps into the error handling logic in the framework. It may work fine in short term but will be certainly error-prone.

**Comment by dumbledad** [comment]
Thanks @troydai, good luck with the fix. If you can point me to exemplar code showing where I should be setting the 401 instead of in the exception handling that would be great too.

**Comment by troydai** [comment]
@dumbledad can you try some of the solutions provided in issue #699 ? Feel free to let me know if none of them works for you.

**Comment by Tratcher** [comment]
Yeah, throwing exceptions from the events is not a recommended way to do control flow. Instead, set the status code inside the event and call AuthenticationFailedContext.HandleResponse().

**Comment by troydai** [comment]
Addressed in 0314632696464efaa9d34d08233e0a292336361d

**Comment by dumbledad** [comment]
The solution in [issue #699](https://github.com/aspnet/Security/issues/699) involves adding this code:


```csharp
OnAuthenticationFailed = context =>
{
    context.Response.StatusCode = StatusCodes.Status401Unauthorized;
    context.AuthenticationTicket = new Microsoft.AspNet.Authentication.AuthenticationTicket(new ClaimsPrincipal(), new Microsoft.AspNet.Http.Authentication.AuthenticationProperties(), string.Empty);
    context.HandleResponse();
    return Task.FromResult(0);
}
```

My existing code that I'm seeking to replace is in `Startup.Configure(IApplicationBuilder app, IHostingEnvironment env)`. Where does this replacement code go?
